### PR TITLE
Don't offer "introduce local constant" for constants

### LINKED ISSUE #47772: Don't offer "introduce local constant" for local constants
This seems redundant:
image

The same happens for non-local constants that are already constants:
image

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Fixes #47772

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 1]
```diff
@@ -7813,5 +7813,74 @@ void M()
...
+class C
+{
+    const int foo = ([|10|]);
+}
+");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceVariable)]
+        [WorkItem(47772, "https://github.com/dotnet/roslyn/issues/47772")]
+        public async Task DonotIntroduceConstantForConstant_NotForSubExpression()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:7861]: this name is just confusing.  having double negatives is not a good thing in names :)

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 2]
```diff
@@ -7813,5 +7813,74 @@ void M()
...
+    {
+        const int foo = [|10|];
+    }
+}
+");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceVariable)]
+        [WorkItem(47772, "https://github.com/dotnet/roslyn/issues/47772")]
+        public async Task DonotIntroduceConstantForConstant_Member()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:7835]:
```suggestion
        public async Task DoNotIntroduceConstantForConstant_Member()
```

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 3]
```diff
@@ -7813,5 +7813,74 @@ void M()
     }
 }");
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceVariable)]
+        [WorkItem(47772, "https://github.com/dotnet/roslyn/issues/47772")]
+        public async Task DonotIntroduceConstantForConstant_Local()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:7819]:
```suggestion
        public async Task DoNotIntroduceConstantForConstant_Local()
```

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs [hunk 4]
```diff
@@ -100,6 +100,13 @@ public State(TService service, SemanticDocument document)
                 IsConstant = Document.SemanticModel.GetConstantValue(Expression, cancellationToken).HasValue;
 
                 // Note: the ordering of these clauses are important.  They go, generally, from
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs:102]: you split this comment in half :)

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService.cs [hunk 5]
```diff
@@ -28,6 +28,24 @@ public CSharpIntroduceVariableService()
...
+                        Parent: VariableDeclaratorSyntax declarator
+                    }
+                })
+            {
+                // Sub-expression
+                return false;
+            }
+
+            return declarator.GetAncestorOrThis<LocalDeclarationStatementSyntax>()?.IsConst == true
+                || declarator.GetAncestorOrThis<FieldDeclarationSyntax>()?.Modifiers.Any(SyntaxKind.ConstKeyword) == true;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService.cs:46]: so i think all this code can be moved into the base class.  from there you can use ISyntaxFacts and/or SyntaxGenerator.  Effectively, you just want to get get the parent of the expr, see if you're in a var-decl, and if so, grab hte modifiers from it (i.e. SyntaxGenerator.GetModifiers).   Then you can just check if the decl is const and  bail.

Comment (huoyaoyuan) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService.cs:46]: I can't find a handful way to check if current node is the body of variable initializer, and get the corresponding variable. Even the IOperation tree of two languages are not identical.

And also, this refactoring providing hasn't been using ISyntaxFacts and related.

Comment (huoyaoyuan) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService.cs:46]: @CyrusNajmabadi ping for my PRs assigned to you

**Comment by huoyaoyuan** [comment]
Looks like a flaky test.

**Comment by CyrusNajmabadi** [comment]
Looking at this today.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by huoyaoyuan** [comment]
@CyrusNajmabadi Needs someone other to merge this?

**Comment by CyrusNajmabadi** [comment]
Nope!  Just fell off my radar.  Thanks!

**Pull request merged**